### DISCUSSION TITLE
Tracing results in "TypeError: cannot read property 'internal' of null

### DISCUSSION THREAD

**Discussion by PeteSpencer** [vfsfitvnm/frida-il2cpp-bridge#405] [Q&A] [answered]
Hello,

I am just starting to explore an game with dnSpy and frida. 
My problem is that if I try to get a trace into a class i got that error:
`TypeError: cannot read property 'internal' of null`
I12Cpp.trace(true).classes(ClassName).and().attach()

I can get into it with hooking the class and methods and I tried to debug there, because it seems the only working method in this app. 

Is this because the target app is protected against tracing? Or did I something wrong?

I am using the latest frida and latest libs.

Some ideas? 

Best regards
Pete

**Comment by PeteSpencer** [comment]
I tried it with another game. And it works. :(
Is there a way to trace protected apps?

**Comment by vfsfitvnm** [comment]
Did you wrap your code inside a `Il2Cpp.perform`?

**Reply by vfsfitvnm** [reply]
Well, I guess so...
I believe that game altered their exports so that `il2cpp_thread_get_current` returns null

**Reply by vfsfitvnm** [reply]
Would you try
```ts
Il2Cpp.perform(() => {
    console.log(Il2Cpp.currentThread);
});
```

**Reply by PeteSpencer** [reply]
Thanks for your fast answer. I will try it when I am home and will give you more information about the case. <3 I does not expect an answer today xD

**Reply by vfsfitvnm** [reply]
If that snippet prints null, you can override `il2cpp_thread_current` as follows:
```ts
import "frida-il2cpp-bridge";

const il2cpp_thread_current_impl = new NativeCallback(
    () => Il2Cpp.corlib.class("System.Threading.Thread").method<Il2Cpp.Object>("GetCurrentThread").invoke().handle,
    "pointer",
    []
);

(globalThis as any).IL2CPP_EXPORTS = {
    il2cpp_thread_current: () => il2cpp_thread_current_impl
};

Il2Cpp.perform(() => {

});
```

**Reply by PeteSpencer** [reply]
Still got this. With and without your last added snippet.

```console 
TypeError: cannot read property 'internal' of null
starting main..
main loaded...
{
  "handle": "0x6d4b74a480"
}
```

```javascript
Java.perform(async () => {
  await ensureModulesInitialized(...modulesList);
  il2cpp = Module.findBaseAddress('libil2cpp.so') as NativePointer;
  console.log('starting main..');
  Il2Cpp.perform(() => {
    // this changes everytime I deploy the script
    console.log('current Thread', JSON.stringify(Il2Cpp.currentThread, null, 2));
  });
  main().catch((e) => console.error(e));
});
```

my main

```javascript
async function main() {
  console.log('main loaded...');

  const MainActivity = await getActivity(APP_MAIN_ACTIVITY);
  if (!MainActivity) throw new Error('MainActivity not found!');

  // Getting unity classes
  const AssemblyCSharp = Il2Cpp.domain.assembly('Assembly-CSharp').image;

  const PlayerState= AssemblyCSharp.class('com.the.class.PlayerState');

  Il2Cpp.trace(true).classes(PlayerState).and().attach();
}
```

Does the handle of the thread should change everytime the script is re-executed?

**Reply by PeteSpencer** [reply]
The strange thing is, that hooking the methods of the class working fine.

**Reply by vfsfitvnm** [reply, marked as answer]
Uhm, `main` isn't executed inside a `Il2Cpp.perform` block

**Reply by PeteSpencer** [reply]
ohh i see

**Reply by PeteSpencer** [reply]
Ohhh it works!!!
I just didn't see it... I have not seen Java.perform and Il2cpp.perform...

But why does it work with a different game?

Thank you being patient with me <3 ♡

**Reply by vfsfitvnm** [reply]
Perhaps `main` is executed before the callback passed to `Il2Cpp.perform` is done (for some reason), hence when `main` is executed, the current thread is still attached to IL2CPP

**Reply by vfsfitvnm** [reply]
FYI: https://github.com/vfsfitvnm/frida-il2cpp-bridge/wiki/Changelog#v081 (take a look at "▶ New threading model")

**Reply by PeteSpencer** [reply]
Now it looks like I got blocked by something^^

```console
Error: abort was called
    at invokeRaw (il2cpp/structs/method.ts:233)
    at toString (il2cpp/structs/object.ts:58)
    at concat (native)
    at <anonymous> (il2cpp/tracer.ts:297)
    at map (native)
    at callback (il2cpp/tracer.ts:297)
Error: abort was called
    at invokeRaw (il2cpp/structs/method.ts:233)
    at toString (il2cpp/structs/object.ts:58)
    at concat (native)
    at <anonymous> (il2cpp/tracer.ts:297)
    at map (native)
    at callback (il2cpp/tracer.ts:297)
Process terminated
```

**Reply by vfsfitvnm** [reply]
https://github.com/vfsfitvnm/frida-il2cpp-bridge/issues/282
